### PR TITLE
Multilingual Librispeech ASR2 + ASR1 baselines

### PR DESCRIPTION
**Pull request by juice500ml**
## What?
- ASR2 recipe for the Multilingual Librispeech
- Training from scratch, 10h split (10h per language, total 8 language)
- ASR2, ASR1 Fbank baseline, ASR1 SSL baseline

## Why?
- MLS experiments for the ASR2 paper

## See also
- ASR1 refactoring PR must be finished first: https://github.com/espnet/espnet/pull/5323

### CODE REVIEW

**Comment by sw005320** [comment]
Is it ready for review?
If so, please change the status from draft to ready for review.

**Comment by juice500ml** [comment]
@sw005320 This PR (https://github.com/espnet/espnet/pull/5323) is a precursor of this PR, so it'd be better if we fix that first. Let me try to fix the previous PRs' tests first, will try to fix it this week 😄

File: egs2/mls/asr2/run.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    --speed_perturb_factors "" \
+    --asr_config "${asr_config}" \
+    --inference_config "${inference_config}" \
+    --train_set "${train_set}" \
+    --valid_set "${valid_set}" \
+    --test_sets "${test_sets}" \
+    "$@"
+    # --src_bpe_train_text "data/${train_set}/text.${src_case}.${src_lang}" \
+    # --tgt_bpe_train_text "data/${train_set}/text.${tgt_case}.${tgt_lang}" \
+    # --lm_train_text "data/${train_set}/text.${tgt_case}.${tgt_lang} data/local/other_text/text" \
```
Comment (ftshijt) [inline egs2/mls/asr2/run.sh:64]:
```suggestion
```

File: egs2/mls/asr2/run.sh [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    --ngpu 1 \
+    --src_lang ${src_lang} \
+    --tgt_lang ${tgt_lang} \
+    --src_token_type "bpe" \
+    --src_nbpe $src_nbpe \
+    --tgt_token_type "bpe" \
+    --tgt_nbpe $tgt_nbpe \
+    --src_case ${src_case} \
+    --tgt_case ${tgt_case} \
+    --speed_perturb_factors "" \
```
Comment (ftshijt) [inline egs2/mls/asr2/run.sh:55]:
```suggestion
```

File: egs2/mls/asr2/RESULTS.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# RESULTS
+## Environments
+- date: `Wed Sep 13 18:26:42 EDT 2023`
+- python version: `3.8.6 (default, Dec 17 2020, 16:57:01)  [GCC 10.2.0]`
+- espnet version: `espnet 202308`
+- pytorch version: `pytorch 1.13.1+cu117`
+- Git hash: `a8bc43b1bfc9518da7dd8be4cad0ef346ef222fc`
+  - Commit date: `Sun Aug 20 16:17:23 2023 -0400`
+
+## exp/asr_smallerbatch_wamrup10k_lr0.0001_e200_raw_wavlm_large_21_full_km1000_bpe_rm3000_bpe_ts150
```
Comment (ftshijt) [inline egs2/mls/asr2/RESULTS.md:11]: You may consider put the model to huggingface as well

File: egs2/mls/asr2/cmd.sh [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+# If jobs failed, your configuration might be wrong for your environment.
+#
+#
+# The official documentation for run.pl, queue.pl, slurm.pl, and ssh.pl:
+#   "Parallelization in Kaldi": http://kaldi-asr.org/doc/queue.html
+# =========================================================~
+
+
+# Select the backend used by run.sh from "local", "stdout", "sge", "slurm", or "ssh"
+cmd_backend='slurm'
```
Comment (ftshijt) [inline egs2/mls/asr2/cmd.sh:31]:
```suggestion
cmd_backend='local'
```

File: egs2/mls/asr2/conf/slurm.conf [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# Default configuration
```
Comment (ftshijt) [inline egs2/mls/asr2/conf/slurm.conf:1]: please set it as default (our CI check would ask for that

File: egs2/mls/asr2/run.sh [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+kmeans_feature=wavlm_large/21  # use model_type/layer_index
+nclusters=1000
+
+src_lang=$(echo "${kmeans_feature}_full_km${nclusters}" | tr "/" "_")
+tgt_lang=en
```
Comment (ftshijt) [inline egs2/mls/asr2/run.sh:12]: Since the target language is not multilingual anymore, you may consider change it to `multi` or `multilingual` (the tag is mostly serving for token naming etc, so please feel free to change)

File: egs2/mls/asr2/run.sh [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    done
+else
+    test_sets="mls_${lang}_test"
+fi
+
+asr_config=conf/train_discrete_asr_e_branchformer1.yaml
+inference_config=conf/decode_ctc0.3.yaml
+
+src_nbpe=3000
+tgt_nbpe=150
```
Comment (ftshijt) [inline egs2/mls/asr2/run.sh:33]: maybe we should also consider using different bpe depending on the size of the data. I feel the current setting would be good for 1h, 10h, or single languages, but definitely not good for all the data

File: egs2/mls/asr2/run.sh [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+# ts: true sequence
+# rm: deduplicated sequence which removes duplicated tokens
+src_case="rm"
+tgt_case="ts"
+
+./asr2.sh \
+    --local_data_opts "--lang ${lang} --data_split ${data_split}" \
+    --portion 1.0 \
```
Comment (ftshijt) [inline egs2/mls/asr2/run.sh:42]: ditto, portion 1.0 is too large for the whole set

**Review by ftshijt — COMMENTED** [review]
Thanks for the contribution! Some minor comments regarding configurations and formating

File: .gitignore [hunk 9]
```diff
@@ -18,7 +18,9 @@ check_autopep8
 .coverage
 htmlcov
 coverage.xml*
-bats-core/
+test_utils/bats-core/
```
Comment (simpleoier) [inline .gitignore:21]: Not sure if we are supposed to touch this file in general PRs.

File: egs2/TEMPLATE/asr2/asr2.sh [hunk 10]
```diff
@@ -680,6 +680,12 @@ fi
 if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ] && ! [[ " ${skip_stages} " =~ [[:space:]]4[[:space:]] ]]; then
     log "Stage 4a: Perform Kmeans using ${kmeans_feature_type} features"
 
+    if [ ${ngpu} -gt 0 ]; then
```
Comment (simpleoier) [inline egs2/TEMPLATE/asr2/asr2.sh:683]: Sorry, I don't want to mix `ngpu` (usually used in training LM / ASR models) with `use_gpu` here.

File: egs2/mls/asr2/cmd.sh [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+elif [ "${cmd_backend}" = slurm ]; then
+    # The default setting is written in conf/slurm.conf.
+    # You must change "-p cpu" and "-p gpu" for the "partition" for your environment.
+    # To know the "partion" names, type "sinfo".
+    # You can use "--gpu * " by default for slurm and it is interpreted as "--gres gpu:*"
+    # The devices are allocated exclusively using "${CUDA_VISIBLE_DEVICES}".
+
+    export train_cmd="slurm.pl"
+    export cuda_cmd="slurm.pl"
+    export decode_cmd="slurm.pl --num_threads 4 --mem 2000M"
```
Comment (simpleoier) [inline egs2/mls/asr2/cmd.sh:85]: You may clean your customized settings from this file.

File: utils/simple_dict.sh
Comment (simpleoier) [inline utils/simple_dict.sh:1]: how about moving this file into `scripts`?

**Review by simpleoier — COMMENTED** [review]
Thanks for your efforts! I left some comments.

File: egs2/mls/asr2/run.sh (hunk 1 above)
Comment (juice500ml) [reply egs2/mls/asr2/run.sh]: Applied in https://github.com/espnet/espnet/pull/5441/commits/91c2b06069b95492b9f6eee9ffb5e1cc84bdb896 Thanks 👍

File: egs2/mls/asr2/run.sh (hunk 2 above)
Comment (juice500ml) [reply egs2/mls/asr2/run.sh]: Applied in https://github.com/espnet/espnet/pull/5441/commits/91c2b06069b95492b9f6eee9ffb5e1cc84bdb896 Thanks 👍

File: egs2/mls/asr2/conf/slurm.conf (hunk 5 above)
Comment (juice500ml) [reply egs2/mls/asr2/conf/slurm.conf:1]: Applied in https://github.com/espnet/espnet/pull/5441/commits/91c2b06069b95492b9f6eee9ffb5e1cc84bdb896 Thanks 👍

File: egs2/mls/asr2/run.sh (hunk 6 above)
Comment (juice500ml) [reply egs2/mls/asr2/run.sh]: Applied in https://github.com/espnet/espnet/pull/5441/commits/91c2b06069b95492b9f6eee9ffb5e1cc84bdb896 Thanks 👍

File: egs2/mls/asr2/run.sh (hunk 7 above)
Comment (juice500ml) [reply egs2/mls/asr2/run.sh:33]: Great point! I set the default as 10h in https://github.com/espnet/espnet/pull/5441/commits/91c2b06069b95492b9f6eee9ffb5e1cc84bdb896
I don't think we want to do the full dataset for now.

File: egs2/mls/asr2/run.sh (hunk 8 above)
Comment (juice500ml) [reply egs2/mls/asr2/run.sh:42]: Similar with the above, set the default to 10h!

File: egs2/TEMPLATE/asr2/asr2.sh (hunk 10 above)
Comment (juice500ml) [reply egs2/TEMPLATE/asr2/asr2.sh]: Added `gpu_kmeans` parameter (following the `gpu_inference` style variable naming), and set it into default in https://github.com/espnet/espnet/pull/5441/commits/91c2b06069b95492b9f6eee9ffb5e1cc84bdb896 Thanks for pointing out!

File: egs2/mls/asr2/cmd.sh (hunk 11 above)
Comment (juice500ml) [reply egs2/mls/asr2/cmd.sh]: Applied in https://github.com/espnet/espnet/commit/91c2b06069b95492b9f6eee9ffb5e1cc84bdb896 Thanks 👍

**Comment by juice500ml** [comment]
Just found out that `fr` evaluation was killed during mid-eval... Need to run eval again and update the README.md. I'll come back to this after eval is done 🥲

File: utils/simple_dict.sh
Comment (juice500ml) [reply utils/simple_dict.sh:1]: Applied in [29d7cb8](https://github.com/espnet/espnet/pull/5441/commits/29d7cb8453486b9073f729866a8cb3d4a8c203bb), thx!
(Related discussion: https://github.com/espnet/espnet/pull/5323#discussion_r1298928177)

File: egs2/mls/asr2/RESULTS.md (hunk 3 above)
Comment (juice500ml) [reply egs2/mls/asr2/RESULTS.md]: Uploaded to huggingface (5584524), thx for the tip!

**Comment by juice500ml** [comment]
I tried to apply all the great reviews to the PR, and it seems that now it is close to merging!

File: egs2/mls/asr2/RESULTS.md
Comment (sw005320) [inline egs2/mls/asr2/RESULTS.md:1]: I observed degradations for some languages from fbank.
Can you summarize them here?
Will you plan to tune it more?

**Comment by sw005320** [comment]
I think this PR is almost there.

File: egs2/mls/asr2/run.sh (hunk 8 above)
Comment (simpleoier) [reply egs2/mls/asr2/run.sh:42]: Do you think it is a good idea to have a "comment" to explain the portion `1.0`?

File: egs2/mls/asr2/RESULTS.md
Comment (juice500ml) [reply egs2/mls/asr2/RESULTS.md:1]: CER comparison

| Model | ASR1 FBANK | ASR1 SSL | ASR2 |
|--------|--------|--------|--------|
| EN | 22.1 | 15.3 | 12.4 |
| ES | 7.0 | 7.0 | 7.9 |
| DE | 10.3 | 10.4 | 11.9 |
| FR | 13.5 | 13.6 | 17.0 |
| IT | 7.1 | 6.9 | 7.9 |
| NL | 11.4 | 10.8 | 14.6 | 
| PL | 8.2 | 7.2 | 11.3 | 
| PT | 12.7 | 11.8 | 13.2 | 

Currently, all the experiments aren't using the language model (asr1 ssl, asr1 fbank, asr2), so it can be somewhat worse compared to other reported numbers on using MLS 10h (10h per each language, total 8 languages).
Also, for asr1, I tried tuning the learning rate (1e-3, 1e-4, 1e-5) and chose the best among three. In comparison, for asr2, I tuned learning rate, batch size, kmeans k, source bpe, and target bpe. I think, to increase the asr2 performance, we need some modifications on the algorithm.
Interestingly, even though asr2 is generally a bit worse, it shows much better performance on English while showing limited performance on French. I'm suspecting that underlying ssl model is English-friendly so asr1 ssl and asr2 is somewhat better than asr1 fbank, but I'm not so sure.
For now, I wasn't thinking about tuning it more (because ES performance is similar to the original ES-only model's performance with CER=7.1), but if you think it's necessary, I'm open to more tuning.

**Comment by sw005320** [comment]
@simpleoier and/or @ftshijt, if it is OK for you, please merge this PR.
